nilearn/input_data/nifti_spheres_masker.py

```diff
@@ -85,6 +86,94 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
     return X, A
 
 
+def _get_spheres_rows(seeds, radius, allow_overlap, mask_img):
+    '''Utility function to get only the rows which are occupied by sphere at
+    given seed locations and the provided radius. Rows are in target_affine and 
+    target_shape space.
+
+    Parameters
+    ----------
+    seeds_: List of triplets of coordinates in native space
+        Seed definitions. List of coordinates of the seeds in the same space
+        as target_affine.
+    radius: float
+        Indicates, in millimeters, the radius for the sphere around the seed.
+    allow_overlap: boolean
+        If False, an error is raised if the maps overlaps (ie at least two
+        maps have a non-zero value for the same voxel).
+    mask_img: Niimg-like object,
+        Mask as a reference space to project the spheres back into brain space.
+
+    Returns
+    -------
+    spheres_map: list of lists 
+        The spheres' location (flattened array), one list for each sphere.
+    '''
+    seeds = list(seeds)
+
+    # Load masking data
+    mask, affine = masking._load_mask_img(mask_img)
+
+    # Get coordinate for alle voxels inside of mask
+    brain_coords = list(np.ndindex(*mask.shape))
+
+    # Get the closest voxels for each seed
+    nearests = []
+    for sx, sy, sz in seeds:
+        nearest = np.round(coord_transform(sx, sy, sz, 
+                           np.linalg.inv(affine)))
+        nearest = nearest.astype(int)
+        nearest = (nearest[0], nearest[1], nearest[2])
+        try:
+            nearests.append(brain_coords.index(nearest))
+        except ValueError:
+            nearests.append(None)
+
+    brain_coords = np.asarray(list(zip(*brain_coords)))
+    # Calculate mask and the corresponding index in flattended array.
+    mask_coords = list(zip(*np.where(mask == 1)))
+    mask_coords = np.asarray(mask_coords).astype(int).T
+    mask_idx = np.vstack([np.in1d(brain_coords[0, :], mask_coords[0, :]),
+                          np.in1d(brain_coords[1, :], mask_coords[1, :]),
+                          np.in1d(brain_coords[2, :], mask_coords[2, :])]
+                        ).all(0)
+
+    # Bring brain_coords into the masks space.
+    brain_coords = coord_transform(brain_coords[0], brain_coords[1],
+                                  brain_coords[2], affine)
+    brain_coords = np.asarray(brain_coords).T
+    # Calculate the adjacent coordinates in radius.
+    clf = neighbors.NearestNeighbors(radius=radius)
+    adjacency = clf.fit(brain_coords).radius_neighbors_graph(seeds)
```

Comment:
that takes time but we expect it to, it is more or less the bulk of the function's work